src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
+            throw new IndexOutOfBoundsException("Can't add element on index " + index);
+        }
+        if (size == 0 && index == 0) {
+            Node<T> node = new Node<>(tail, value, null);
+            if (head == null) {
+                head = node;
+                tail = node;
+            } else {
+                tail.next = node;
+                tail = node;
+            }
+            size++;
+            return;
+        }
+        if (size != 0 && index == 0) {
+            Node<T> node = new Node<>(null, value, head);
+            head.prev = node;
+            head = node;
+            size++;
+            return;
+        }
+        if (size != 0 && index == size) {
+            Node<T> node = new Node<>(tail, value, null);
+            tail.next = node;
+            tail = node;
+            size++;
+            return;
+        }
+        Node<T> addedNode = findElementByIndex(index);
+        Node<T> node = new Node<>(addedNode.prev, value, addedNode);
+        addedNode.prev.next = node;
+        addedNode.prev = node;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T addedElement : list) {
+            add(addedElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (checkElementByIndex(index)) {
+            throw new IndexOutOfBoundsException("Element with index " + index + " does not exist");
+        }
+        return findElementByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (checkElementByIndex(index)) {
+            throw new IndexOutOfBoundsException("Can't set element on index " + index);
+        }
+        Node<T> replacedNode = findElementByIndex(index);
+        T valueReturned = replacedNode.value;
+        replacedNode.value = value;
+        return valueReturned;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (checkElementByIndex(index)) {
+            throw new IndexOutOfBoundsException("Element with index " + index + " does not exist");
+        }
+        Node<T> removedNode = findElementByIndex(index);
+        unlink(removedNode);
+        return removedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            return false;
+        }
+        int counter = 0;
+        Node<T> elementByValue = head;
+        while (!(elementByValue == tail) && !(elementByValue.value == object
+                || elementByValue.value != null
+                && elementByValue.value.equals(object))) {
+            elementByValue = elementByValue.next;
+            counter++;
+        }
+        if (counter < size && (elementByValue.value == object
+                || elementByValue.value != null
+                && elementByValue.value.equals(object))) {
+            unlink(elementByValue);
+            return true;
+        }
```

Comment:
can we simplify all this logic? use while loop here will be enough, you don't need variable `counter`